IML 2025 Assignment 1

Jiwoo Hong m12445507  
Yeongjae Park m12446610  
Yoonhyeok Lee m12448710  

## **README**
- **The training and testing will be automatically done if you press 'Run All' button in anaconda jupyter notebook or visual studio code.**  
- **The final result for training and testing will be displayed at the end of this notebook file.**  
- **The "Oracle" class will keep track of number of used labels. The numbers of used labels will be displayed with the final result.**
- **Please install sklearn and tqdm and numpy to run this code.**

# Definition of input files

In [19]:
#These filename need to be defined:

#  current dataset directory : data/alldigits.csv
#  amount of data: 6000
#  structure of each datum: [28x28 pixels with value 0x00 ~ 0xFF, 1 label with value 0 ~ 9]

datafile_name = 'data/alldigits.csv'

# Your sections ...

#### Import Classifiers and Utils

In [20]:
from sklearn.svm import SVC
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import GradientBoostingClassifier, RandomForestClassifier
from sklearn.naive_bayes import GaussianNB, MultinomialNB
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score
from tqdm import tqdm

# Calc1: Read data

In [21]:
import numpy as np


all_digits = np.loadtxt(datafile_name, delimiter=',', skiprows=1)



# Preprocessing: Data split

Defined Oracle to 
- Perform uncertainty sampling in training
- Retrieve test data based on active testing
- Keep track of data that the label has been used for training/model selection/testing

In [22]:
import math

DATASET_SIZE = 6000  #  precomputated value

class Oracle:
    def __init__(self, data, train_ratio, validate_ratio, test_ratio):
        self.data = data

        assert train_ratio + validate_ratio + test_ratio == 1.0, "Ratios must sum to 1"
        self.train_ratio = train_ratio
        self.validate_ratio = validate_ratio
        self.test_ratio = test_ratio

        self.train_data = []
        self.validate_data = []
        self.test_data = []

        self.train_used_markers = {}
        self.validate_used_markers = [False] * int(DATASET_SIZE * self.validate_ratio)
        self.test_used_markers = []

        self.surrogate =  MultinomialNB()  # Initialize with a default classifier

        self.X_surrogate = None
        self.y_surrogate = None

    def split_data(self, shuffle=True):
        """
        @param shuffle: If True, shuffle the dataset before splitting.

        @return: Tuple of (train_data, validate_data, test_data)
        Split the dataset into training, validation, and test sets based on the specified ratios.
        """

        x = self.data[:, :-1]
        y = self.data[:, -1]

        if shuffle:
            indices = np.arange(DATASET_SIZE)
            np.random.shuffle(indices)
            x = x[indices]
            y = y[indices]

        self.train_size = int(DATASET_SIZE * self.train_ratio)
        self.validate_size = int(DATASET_SIZE * self.validate_ratio)
        self.test_size = DATASET_SIZE - self.train_size - self.validate_size

        self.train_data = (x[:self.train_size], y[:self.train_size])
        self.validate_data = (x[self.train_size:self.train_size + self.validate_size], y[self.train_size:self.train_size + self.validate_size])
        self.test_data = (x[self.train_size + self.validate_size:], y[self.train_size + self.validate_size:])

        # Markers for used data
        self.train_used_markers = {}
        self.validate_used_markers = [False] * self.validate_size
        self.test_used_markers = [False] * self.test_size

        return self.train_data, self.validate_data, self.test_data

    def retrieve_uncertain_training_data(self, probabilty_matrix, classifier_name, num_samples=1):

        # Calculate the uncertainty for each data point
        uncertainty = np.zeros(probabilty_matrix.shape[0])
        for i in range(probabilty_matrix.shape[0]):
            uncertainty[i] = self.entropy(probabilty_matrix[i])

        # Sort the data points by uncertainty
        uncertain_indices = np.argsort(uncertainty)[::-1]
        uncertain_data = []
        uncertain_labels = []
        for i in range(num_samples):
            index = uncertain_indices[i]
            uncertain_data.append(self.train_data[0][index])
            uncertain_labels.append(self.train_data[1][index])
            self.train_used_markers[classifier_name][index] = True

        return np.array(uncertain_data), np.array(uncertain_labels)

    def retrieve_initial_training_data(self, classifier_name):
        if classifier_name not in self.train_used_markers.keys():
            self.train_used_markers[classifier_name] = [False] * self.train_size

        # get each data of all labels
        initial_data = []
        initial_labels = []
        for label in range(10):
            label_indices = np.where(self.train_data[1] == label)[0]
            if len(label_indices) > 0:
                initial_index = label_indices[0]
                initial_data.append(self.train_data[0][initial_index])
                initial_labels.append(self.train_data[1][initial_index])
                self.train_used_markers[classifier_name][initial_index] = True

        return np.array(initial_data), np.array(initial_labels)

    def train_surrogate_model(self, classifier_name):
        # Find trained X and concatenate them
        if classifier_name not in self.train_used_markers.keys():
            raise ValueError(f"Classifier '{classifier_name}' has not been used to mark training data.")

        used_X_train = np.vstack([self.train_data[0][i] for i in range(self.train_size) if self.train_used_markers[classifier_name][i]])
        used_y_train = np.hstack([self.train_data[1][i] for i in range(self.train_size) if self.train_used_markers[classifier_name][i]])

        if sum(self.test_used_markers) == 0:
            X_surrogate = used_X_train
            y_surrogate = used_y_train
        
        else:
            X_surrogate = self.X_surrogate
            y_surrogate = self.y_surrogate

        self.surrogate.partial_fit(X_surrogate, y_surrogate, classes=np.unique(self.train_data[1]))

    def entropy(self, probabilty_vector):
        return -np.sum(probabilty_vector * np.log(probabilty_vector + 1e-10))  # Adding a small value to avoid log(0)

    def retrieve_validation_data(self, num_samples=1):
        self.validate_used_markers[0: num_samples] = [True] * num_samples

        return self.validate_data[0][0: num_samples], self.validate_data[1][0: 0 + num_samples]

    def retrieve_test_data(self, probability_matrix, num_samples=1):
        # Calculate aqcuisition function
        y_star = np.argmax(probability_matrix, axis=1)
        q = np.zeros(self.test_data[0].shape[0])

        # Calculate aqcuisition function value for each index with resprect to accuracy
        # Only test data that are not observed will be sampled
        for i in range(self.test_data[0].shape[0]):
            q[i] = 0 if self.test_used_markers[i] else 1 - self.surrogate.predict_proba(self.test_data[0][i].reshape(1, -1))[0][int(y_star[i])] + 1e-13

        # Normalize all q_i to make them as a sampling probabilities
        q = q / sum(q)

        uncertain_data = []
        uncertain_labels = []

        self.X_surrogate = []
        self.y_surrogate = []

        for _ in range(num_samples):
            index = np.random.choice(self.test_data[0].shape[0], p=q)
            uncertain_data.append(self.test_data[0][index])
            uncertain_labels.append(self.test_data[1][index])
            self.test_used_markers[index] = True

            q[index] = 0  # Set the chosen index to 0 to avoid re-selection
            q = q / sum(q)  # Normalize again after removing the selected index

            self.X_surrogate.append(self.test_data[0][index])
            self.y_surrogate.append(self.test_data[1][index])

        self.X_surrogate = np.array(self.X_surrogate).reshape(-1, self.test_data[0].shape[1])
        self.y_surrogate = np.array(self.y_surrogate)

        return np.array(uncertain_data), np.array(uncertain_labels)

    def count_used_data(self):

        merge_train_used_markers = [False] * self.train_size
        # OR operate for all classifiers in train_used_markers
        for model in self.train_used_markers.keys():
            merge_train_used_markers = [merge_train_used_markers[i] or self.train_used_markers[model][i] for i in range(self.train_size)]

        train_used_count = sum(merge_train_used_markers)
        validate_used_count = sum(self.validate_used_markers)
        test_used_count = sum(self.test_used_markers)

        return train_used_count, validate_used_count, test_used_count


# Explore

#### Initialize Oracle

In [23]:
oracle = Oracle(
    data=all_digits,
    train_ratio=0.75,
    validate_ratio=0.05,
    test_ratio=0.2
)

train_data, validate_data, test_data = oracle.split_data(shuffle=True)
count_of_initial_data = 10

#### Validation results for each classifier

In [24]:
val_acc = {}

Use Active Learning using uncentainty sampling  
~~Naive Bayes~~, Multinomial Bayes, SVM, ~~Logistic Regression~~, Gradient Boosting, Random Forest, ~~Decision Tree~~ were trained  
All Classifiers sample batch with size of 10 because of training time.

In [25]:
models = {}

# ~~Calc2: Naive Bayes~~

In [26]:
"""
TRAIN_ITERATIONS = 1200  #  precomputated value
SAMPLE_BATCH_SIZE = 10

initial_data, initial_labels = oracle.retrieve_initial_training_data("nb")
X_train, y_train = initial_data, initial_labels

X_val, y_val = oracle.retrieve_validation_data(num_samples=300)

pbar = tqdm(range((TRAIN_ITERATIONS - count_of_initial_data) // SAMPLE_BATCH_SIZE), desc="training iterations")

gnb = GaussianNB().partial_fit(X_train, y_train, classes=np.unique(y_train))

for i in pbar:

    # Get the probabilty predictions for the training data
    gnb_pred_proba = gnb.predict_proba(train_data[0])

    # Retrieve uncertain data
    uncertain_data, uncertain_labels = oracle.retrieve_uncertain_training_data(gnb_pred_proba, "nb", num_samples=SAMPLE_BATCH_SIZE)

    # Add the uncertain data to the training set
    X_train = uncertain_data
    y_train = uncertain_labels

    # Train the gaussian Naive Bayes classifier
    gnb.partial_fit(X_train, y_train)

models['nb'] = gnb

y_pred_train = gnb.predict(X_train)
accuracy = accuracy_score(y_train, y_pred_train)

y_pred_val = gnb.predict(X_val)
val_accuracy = accuracy_score(y_val, y_pred_val)

val_acc['nb'] = val_accuracy
print(f"Train accuracy: {accuracy * 100:.2f}%")
print(f"Validation accuracy: {val_accuracy * 100:.2f}%")
"""

'\nTRAIN_ITERATIONS = 1200  #  precomputated value\nSAMPLE_BATCH_SIZE = 10\n\ninitial_data, initial_labels = oracle.retrieve_initial_training_data("nb")\nX_train, y_train = initial_data, initial_labels\n\nX_val, y_val = oracle.retrieve_validation_data(num_samples=300)\n\npbar = tqdm(range((TRAIN_ITERATIONS - count_of_initial_data) // SAMPLE_BATCH_SIZE), desc="training iterations")\n\ngnb = GaussianNB().partial_fit(X_train, y_train, classes=np.unique(y_train))\n\nfor i in pbar:\n\n    # Get the probabilty predictions for the training data\n    gnb_pred_proba = gnb.predict_proba(train_data[0])\n\n    # Retrieve uncertain data\n    uncertain_data, uncertain_labels = oracle.retrieve_uncertain_training_data(gnb_pred_proba, "nb", num_samples=SAMPLE_BATCH_SIZE)\n\n    # Add the uncertain data to the training set\n    X_train = uncertain_data\n    y_train = uncertain_labels\n\n    # Train the gaussian Naive Bayes classifier\n    gnb.partial_fit(X_train, y_train)\n\nmodels[\'nb\'] = gnb\n\ny_pr

# Calc2: Multinomial Naive Bayes

In [27]:
TRAIN_ITERATIONS = 1200
SAMPLE_BATCH_SIZE = 10

initial_data, initial_labels = oracle.retrieve_initial_training_data("mnb")
X_train = initial_data
y_train = initial_labels

X_val, y_val = oracle.retrieve_validation_data(num_samples=300)

mnb = MultinomialNB().partial_fit(X_train, y_train, classes=np.unique(y_train))

pbar = tqdm(range((TRAIN_ITERATIONS - len(X_train)) // SAMPLE_BATCH_SIZE), desc="training iterations")

for i in pbar:

    mnb_pred_proba = mnb.predict_proba(train_data[0])

    uncertain_data, uncertain_labels = oracle.retrieve_uncertain_training_data(
        mnb_pred_proba, "mnb", num_samples=SAMPLE_BATCH_SIZE
    )

    X_train = uncertain_data
    y_train = uncertain_labels

    mnb.partial_fit(X_train, y_train)

models['mnb'] = mnb

y_pred_train = mnb.predict(X_train)
accuracy = accuracy_score(y_train, y_pred_train)

y_pred_val = mnb.predict(X_val)
val_accuracy = accuracy_score(y_val, y_pred_val)
val_acc['mnb'] = val_accuracy

print(f"Train accuracy: {accuracy * 100:.2f}%")
print(f"Validation accuracy: {val_accuracy * 100:.2f}%")

training iterations: 100%|██████████| 119/119 [00:04<00:00, 24.32it/s]

Train accuracy: 60.00%
Validation accuracy: 84.00%


# Calc2: SVM

In [28]:
TRAIN_ITERATIONS = 1200  #  precomputated value
SAMPLE_BATCH_SIZE = 10

initial_data, initial_labels = oracle.retrieve_initial_training_data("svm")
X_train = initial_data
y_train = initial_labels

X_val, y_val = oracle.retrieve_validation_data(num_samples=300)


pbar = tqdm(range((TRAIN_ITERATIONS - count_of_initial_data) // SAMPLE_BATCH_SIZE), desc="training iterations")

for i in pbar:
    # Train the SVM model
    svm = SVC(kernel='linear', C=1, probability=True).fit(X_train, y_train)

    # Get the probability predictions for the training data
    y_pred_proba = svm.predict_proba(train_data[0])

    # Retrieve uncertain data
    uncertain_data, uncertain_label = oracle.retrieve_uncertain_training_data(y_pred_proba, "svm", num_samples=SAMPLE_BATCH_SIZE)

    # Add the uncertain data to the training set
    X_train = np.vstack((X_train, uncertain_data))
    y_train = np.hstack((y_train, uncertain_label))

models['svm'] = svm

y_pred_train = svm.predict(X_train)
accuracy = accuracy_score(y_train, y_pred_train)

y_pred_val = svm.predict(X_val)
val_accuracy = accuracy_score(y_val, y_pred_val)
val_acc['svm'] = val_accuracy

print(f"Train accuracy: {accuracy * 100:.2f}%")
print(f"Validation accuracy: {val_accuracy * 100:.2f}%")

training iterations: 100%|██████████| 119/119 [01:20<00:00,  1.48it/s]


Train accuracy: 99.92%
Validation accuracy: 91.67%


# ~~Calc2: LogisticRegression~~

In [29]:
"""
TRAIN_ITERATIONS = 1200  #  precomputated value
SAMPLE_BATCH_SIZE = 10

initial_data, initial_labels = oracle.retrieve_initial_training_data("lr")
X_train = initial_data
y_train = initial_labels

X_val, y_val = oracle.retrieve_validation_data(num_samples=300)

pbar = tqdm(range((TRAIN_ITERATIONS - count_of_initial_data) // SAMPLE_BATCH_SIZE), desc="training iterations")

for i in pbar:
    # Train the Logistic Regression model
    lr = LogisticRegression(max_iter=2000, n_jobs=-1).fit(X_train, y_train)

    # Get the probability predictions for the training data
    y_pred_proba = lr.predict_proba(train_data[0])

    # Retrieve uncertain data
    uncertain_data, uncertain_label = oracle.retrieve_uncertain_training_data(y_pred_proba, "lr", num_samples=SAMPLE_BATCH_SIZE)

    # Add the uncertain data to the training set
    X_train = np.vstack((X_train, uncertain_data))
    y_train = np.hstack((y_train, uncertain_label))

models['lr'] = lr

y_pred_train = lr.predict(X_train)
accuracy = accuracy_score(y_train, y_pred_train)

y_pred_val = lr.predict(X_val)
val_accuracy = accuracy_score(y_val, y_pred_val)
val_acc['lr'] = val_accuracy

print(f"Train accuracy: {accuracy * 100:.2f}%")
print(f"Validation accuracy: {val_accuracy * 100:.2f}%")
"""

'\nTRAIN_ITERATIONS = 1200  #  precomputated value\nSAMPLE_BATCH_SIZE = 10\n\ninitial_data, initial_labels = oracle.retrieve_initial_training_data("lr")\nX_train = initial_data\ny_train = initial_labels\n\nX_val, y_val = oracle.retrieve_validation_data(num_samples=300)\n\npbar = tqdm(range((TRAIN_ITERATIONS - count_of_initial_data) // SAMPLE_BATCH_SIZE), desc="training iterations")\n\nfor i in pbar:\n    # Train the Logistic Regression model\n    lr = LogisticRegression(max_iter=2000, n_jobs=-1).fit(X_train, y_train)\n\n    # Get the probability predictions for the training data\n    y_pred_proba = lr.predict_proba(train_data[0])\n\n    # Retrieve uncertain data\n    uncertain_data, uncertain_label = oracle.retrieve_uncertain_training_data(y_pred_proba, "lr", num_samples=SAMPLE_BATCH_SIZE)\n\n    # Add the uncertain data to the training set\n    X_train = np.vstack((X_train, uncertain_data))\n    y_train = np.hstack((y_train, uncertain_label))\n\nmodels[\'lr\'] = lr\n\ny_pred_train = l

# Calc2: Gradient Boosting

In [30]:
TRAIN_ITERATIONS = 1200  #  precomputated value
SAMPLE_BATCH_SIZE = 10

initial_data, initial_labels = oracle.retrieve_initial_training_data("gb")
X_train = initial_data
y_train = initial_labels

X_val, y_val = oracle.retrieve_validation_data(num_samples=300)

pbar = tqdm(range((TRAIN_ITERATIONS - count_of_initial_data) // SAMPLE_BATCH_SIZE), desc="training iterations")

for i in pbar:
    # Train the Gradient Boosting model
    gb = GradientBoostingClassifier(
        n_estimators=200,
        learning_rate=0.1,
        max_depth=5,
        subsample=0.8,
        max_features='sqrt',
        min_samples_leaf=3,
    ).fit(X_train, y_train)

    # Get the probability predictions for the training data
    y_pred_proba = gb.predict_proba(train_data[0])

    # Retrieve uncertain data
    uncertain_data, uncertain_label = oracle.retrieve_uncertain_training_data(y_pred_proba, "gb", num_samples=SAMPLE_BATCH_SIZE)

    # Add the uncertain data to the training set
    X_train = np.vstack((X_train, uncertain_data))
    y_train = np.hstack((y_train, uncertain_label))

models['gb'] = gb

y_pred_train = gb.predict(X_train)
accuracy = accuracy_score(y_train, y_pred_train)

y_pred_val = gb.predict(X_val)
val_accuracy = accuracy_score(y_val, y_pred_val)
val_acc['gb'] = val_accuracy

print(f"Train accuracy: {accuracy * 100:.2f}%")
print(f"Validation accuracy: {val_accuracy * 100:.2f}%")

training iterations: 100%|██████████| 119/119 [15:08<00:00,  7.63s/it]

Train accuracy: 99.83%
Validation accuracy: 94.00%


# Calc2: Random Forest

In [31]:
TRAIN_ITERATIONS = 1200  #  precomputated value
SAMPLE_BATCH_SIZE = 10

initial_data, initial_labels = oracle.retrieve_initial_training_data("rf")
X_train = initial_data
y_train = initial_labels

X_val, y_val = oracle.retrieve_validation_data(num_samples=300)

pbar = tqdm(range((TRAIN_ITERATIONS - count_of_initial_data) // SAMPLE_BATCH_SIZE), desc="training iterations")

for i in pbar:
    # Train the Gradient Boosting model
    rf = RandomForestClassifier().fit(X_train, y_train)

    # Get the probability predictions for the training data
    y_pred_proba = rf.predict_proba(train_data[0])

    # Retrieve uncertain data
    uncertain_data, uncertain_label = oracle.retrieve_uncertain_training_data(y_pred_proba, "rf", num_samples=SAMPLE_BATCH_SIZE)

    # Add the uncertain data to the training set
    X_train = np.vstack((X_train, uncertain_data))
    y_train = np.hstack((y_train, uncertain_label))

models['rf'] = rf

y_pred_train = rf.predict(X_train)
accuracy = accuracy_score(y_train, y_pred_train)

y_pred_val = rf.predict(X_val)
val_accuracy = accuracy_score(y_val, y_pred_val)
val_acc['rf'] = val_accuracy

print(f"Train accuracy: {accuracy * 100:.2f}%")
print(f"Validation accuracy: {val_accuracy * 100:.2f}%")


training iterations: 100%|██████████| 119/119 [01:42<00:00,  1.16it/s]

Train accuracy: 99.83%
Validation accuracy: 93.00%


# ~~Calc2: Decision Tree~~

In [32]:
"""
TRAIN_ITERATIONS = 1200  #  precomputated value
SAMPLE_BATCH_SIZE = 10

initial_data, initial_labels = oracle.retrieve_initial_training_data("dt")
X_train = initial_data
y_train = initial_labels

X_val, y_val = oracle.retrieve_validation_data(num_samples=300)

pbar = tqdm(range((TRAIN_ITERATIONS - count_of_initial_data) // SAMPLE_BATCH_SIZE), desc="training iterations")

for i in pbar:
    # Train the Gradient Boosting model
    dt = DecisionTreeClassifier(
        criterion='entropy',
        max_depth=29,
        min_samples_split=4,
        min_samples_leaf=1,
        max_features=None,
    ).fit(X_train, y_train)

    # Get the probability predictions for the training data
    y_pred_proba = dt.predict_proba(train_data[0])

    # Retrieve uncertain data
    uncertain_data, uncertain_label = oracle.retrieve_uncertain_training_data(y_pred_proba, "dt", num_samples=SAMPLE_BATCH_SIZE)

    # Add the uncertain data to the training set
    X_train = np.vstack((X_train, uncertain_data))
    y_train = np.hstack((y_train, uncertain_label))

models['dt'] = dt

y_pred_train = dt.predict(X_train)
accuracy = accuracy_score(y_train, y_pred_train)

y_pred_val = dt.predict(X_val)
val_accuracy = accuracy_score(y_val, y_pred_val)
val_acc['dt'] = val_accuracy

print(f"Train accuracy: {accuracy * 100:.2f}%")
print(f"Validation accuracy: {val_accuracy * 100:.2f}%")
"""

'\nTRAIN_ITERATIONS = 1200  #  precomputated value\nSAMPLE_BATCH_SIZE = 10\n\ninitial_data, initial_labels = oracle.retrieve_initial_training_data("dt")\nX_train = initial_data\ny_train = initial_labels\n\nX_val, y_val = oracle.retrieve_validation_data(num_samples=300)\n\npbar = tqdm(range((TRAIN_ITERATIONS - count_of_initial_data) // SAMPLE_BATCH_SIZE), desc="training iterations")\n\nfor i in pbar:\n    # Train the Gradient Boosting model\n    dt = DecisionTreeClassifier(\n        criterion=\'entropy\',\n        max_depth=29,\n        min_samples_split=4,\n        min_samples_leaf=1,\n        max_features=None,\n    ).fit(X_train, y_train)\n\n    # Get the probability predictions for the training data\n    y_pred_proba = dt.predict_proba(train_data[0])\n\n    # Retrieve uncertain data\n    uncertain_data, uncertain_label = oracle.retrieve_uncertain_training_data(y_pred_proba, "dt", num_samples=SAMPLE_BATCH_SIZE)\n\n    # Add the uncertain data to the training set\n    X_train = np.vst

# Model Selection

Use model that has maximum validate accuracy

In [33]:
for model_name, _val_acc in val_acc.items():
    print(f"Model: {model_name}, Validation Accuracy: {_val_acc * 100:.2f}%")

best_model_name = max(val_acc, key=val_acc.get)
print(f"Best model based on validation accuracy: {best_model_name}")
best_model = models[best_model_name]

Model: mnb, Validation Accuracy: 84.00%
Model: svm, Validation Accuracy: 91.67%
Model: gb, Validation Accuracy: 94.00%
Model: rf, Validation Accuracy: 93.00%
Best model based on validation accuracy: gb


# Active Testing

Used active testing to use minimum data to predict true loss

In [34]:
TEST_ITERATIONS = 800

# Train surrogate
oracle.train_surrogate_model(best_model_name)

X_test = None
y_test = None

# Retrieve test data
for i in tqdm(range(TEST_ITERATIONS), desc="testing iterations"):
    _test_data, _test_label = oracle.retrieve_test_data(best_model.predict_proba(test_data[0]), num_samples=1)

    X_test = _test_data if X_test is None else np.vstack((X_test, _test_data))
    y_test = _test_label if y_test is None else np.hstack((y_test, _test_label))

    oracle.train_surrogate_model(best_model_name)

y_test_pred = best_model.predict(X_test)
accuracy = accuracy_score(y_test, y_test_pred)

print(f"Test accuracy after test data retrieval: {accuracy * 100:.2f}%")

testing iterations:   0%|          | 0/800 [00:00<?, ?it/s]

testing iterations: 100%|██████████| 800/800 [01:59<00:00,  6.68it/s]

Test accuracy after test data retrieval: 93.88%


# ...

In [35]:
#This classifier needs to be defined:
y_test_pred = best_model.predict(X_test)
accuracy = accuracy_score(y_test, y_test_pred)
print(f"Test accuracy after test data retrieval: {accuracy * 100:.2f}%")
# print count of used data
train_used_count, validate_used_count, test_used_count = oracle.count_used_data()
total_used_count = train_used_count + validate_used_count + test_used_count
print(f"train used label: {train_used_count}, validate used label: {validate_used_count}, test used label: {test_used_count}, total used label: {total_used_count}")

Test accuracy after test data retrieval: 93.88%
train used label: 1987, validate used label: 300, test used label: 800, total used label: 3087


# Explanation
## How did you construct the classifier? 
Originally, we made two objects: Oracle and target model. Oracle has all data including validate and test data, but not using itself. Our model, besides, have only training data.

Oracle contains:
* All data including validate and test data
* Surrogate model
* APIs to retrieve train, validate, and test data

Our model contains:
* Only training data
* Classifier

After our model finishes initial training, the model calculate the prediction probability of whole training data and deliver it to oracle.  
Than Oracle calculate its uncertainty by using entropy, and gives back the index of the most uncertain training data. Finally, the model trains with that data.  
While doing uncertainty sampling, we sampled batch of most uncertain data(10 data) to accelerate active learning.  

### What is the accuracy of your classifiers? 
Validation accuracies for the classifiers are:
* Multinomial Naive Bayes: 84.00%
* SVM: 91.67%
* Gradient Boosting: 94.00%
* Random Forest: 93.00%

The best model - Gradient Boosting - achieved 94.00% val accuracy after active learning.

### How did you calculate the accuracy of your classifiers? 
Accuracy is calculated using the `accuracy_score` function from scikit-learn.  
* For validation, predictions on the 300 validation samples are compared to true labels.  
* For testing, predictions on the actively sampled 800 test samples are compared to true labels.

### How confident are you in your accuracy estimate?
We used multinomial NB as a surrogate model. The following are the reasons why we choose miltinomial NB:
* We needed a decorrelated model with our target model, which will be gaussian boosting or random forest.
* Multinomial NB achieved around 80% accuracy at the model validation. In other words, it has a sufficient fidelity.

We have tested over 33 times with the random forest. The following are the mean and varience, and confidence intervals.

$$\mu = 91.22, \sigma = 1.18, n = 33 $$


| Confidence Level | Confidence Interval |
| ---- | ---- |
| 99\% | [90.80, 91.64] |
| 95\% | [90.66, 91.78] |

As shown above, the lower bound of 99% confidence interval exceeds 90.00%.

### How many labels does your algorithm use?
Our algorithm used 3087 out of 6000 possible labels (51%), achieving significant label efficiency compared to using the full dataset.  
(Training: 1987 labels, Validation: 300 labels, Testing: 800 labels) (number of used train labels can vary little due to the random shuffling of dataset)

We are supposed to have used only 1200 train data for a single model, but since we need to train various models and choose one best model, about 800 extra train data are used to train other models.